File: go_backend/internal/domain/psychomatrix.go
package domain

// PsychomatrixWorkingNumbers holds the four derived working numbers of the
// Alexandrov method (see internal/psychomatrix).
type PsychomatrixWorkingNumbers struct {
	First  int `json:"first"`  // W1: sum of all date digits
	Second int `json:"second"` // W2: digital root of W1
	Third  int `json:"third"`  // W3: W1 - 2*firstDigitOfDay
	Fourth int `json:"fourth"` // W4: digital root of abs(W3)
}

// PsychomatrixCell is one of the nine grid cells (digit 1..9). Count is how
// many times the digit appears in the pool; Repeated is the digit rendered
// Count times (e.g. "111") or "" when absent.
type PsychomatrixCell struct {
	Digit    int    `json:"digit"`
	Count    int    `json:"count"`
	Repeated string `json:"repeated"`
	Title    string `json:"title"`
	Meaning  string `json:"meaning"`
}

// PsychomatrixLine is one of the eight lines (rows, columns, diagonals).
// Strength is the total digit count across its three cells.
type PsychomatrixLine struct {
	Key      string `json:"key"`
	Title    string `json:"title"`
	Cells    []int  `json:"cells"`
	Strength int    `json:"strength"`
	Meaning  string `json:"meaning"`
}

// PsychomatrixReading is the full response for GET /api/v1/psychomatrix.
// Cells always holds 9 entries (digit 1..9 in order); Lines always holds 8.
type PsychomatrixReading struct {
	BirthDate      string                     `json:"birth_date"` // YYYY-MM-DD
	WorkingNumbers PsychomatrixWorkingNumbers `json:"working_numbers"`
	Cells          []PsychomatrixCell         `json:"cells"`
	Lines          []PsychomatrixLine         `json:"lines"`
}

File: flutter_app/lib/features/psychomatrix/domain/entities/psychomatrix.dart
import 'package:equatable/equatable.dart';

/// The four derived working numbers of the Alexandrov psychomatrix method.
class PsychomatrixWorkingNumbers extends Equatable {
  const PsychomatrixWorkingNumbers({
    required this.first,
    required this.second,
    required this.third,
    required this.fourth,
  });

  factory PsychomatrixWorkingNumbers.fromJson(Map<String, dynamic> json) {
    return PsychomatrixWorkingNumbers(
      first: (json['first'] as num?)?.toInt() ?? 0,
      second: (json['second'] as num?)?.toInt() ?? 0,
      third: (json['third'] as num?)?.toInt() ?? 0,
      fourth: (json['fourth'] as num?)?.toInt() ?? 0,
    );
  }

  final int first;
  final int second;
  final int third;
  final int fourth;

  @override
  List<Object?> get props => [first, second, third, fourth];
}

/// One of the nine grid cells (digit 1..9) with its occurrence count and text.
class PsychomatrixCell extends Equatable {
  const PsychomatrixCell({
    required this.digit,
    required this.count,
    required this.repeated,
    required this.title,
    required this.meaning,
  });

  factory PsychomatrixCell.fromJson(Map<String, dynamic> json) {
    return PsychomatrixCell(
      digit: (json['digit'] as num?)?.toInt() ?? 0,
      count: (json['count'] as num?)?.toInt() ?? 0,
      repeated: json['repeated'] as String? ?? '',
      title: json['title'] as String? ?? '',
      meaning: json['meaning'] as String? ?? '',
    );
  }

  final int digit;
  final int count;

  /// The digit rendered [count] times, e.g. "111", or "" when absent.
  final String repeated;
  final String title;
  final String meaning;

  bool get isEmpty => count <= 0;

  @override
  List<Object?> get props => [digit, count];
}

/// One of the eight lines (rows, columns, diagonals) of the grid.
class PsychomatrixLine extends Equatable {
  const PsychomatrixLine({
    required this.key,
    required this.title,
    required this.cells,
    required this.strength,
    required this.meaning,
  });

  factory PsychomatrixLine.fromJson(Map<String, dynamic> json) {
    return PsychomatrixLine(
      key: json['key'] as String? ?? '',
      title: json['title'] as String? ?? '',
      cells: ((json['cells'] as List<dynamic>?) ?? const [])
          .map((e) => (e as num?)?.toInt() ?? 0)
          .toList(),
      strength: (json['strength'] as num?)?.toInt() ?? 0,
      meaning: json['meaning'] as String? ?? '',
    );
  }

  final String key;
  final String title;
  final List<int> cells;

  /// Total digit count across the line's three cells.
  final int strength;
  final String meaning;

  @override
  List<Object?> get props => [key, strength];
}

/// The full psychomatrix reading returned by GET /api/v1/psychomatrix.
class PsychomatrixReading extends Equatable {
  const PsychomatrixReading({
    required this.birthDate,
    required this.workingNumbers,
    required this.cells,
    required this.lines,
  });

  factory PsychomatrixReading.fromJson(Map<String, dynamic> json) {
    final cellList = (json['cells'] as List<dynamic>?) ?? const [];
    final lineList = (json['lines'] as List<dynamic>?) ?? const [];
    return PsychomatrixReading(
      birthDate: json['birth_date'] as String? ?? '',
      workingNumbers: PsychomatrixWorkingNumbers.fromJson(
        (json['working_numbers'] as Map<String, dynamic>?) ?? const {},
      ),
      cells: cellList
          .whereType<Map<String, dynamic>>()
          .map(PsychomatrixCell.fromJson)
          .toList(),
      lines: lineList
          .whereType<Map<String, dynamic>>()
          .map(PsychomatrixLine.fromJson)
          .toList(),
    );
  }

  final String birthDate;
  final PsychomatrixWorkingNumbers workingNumbers;
  final List<PsychomatrixCell> cells;
  final List<PsychomatrixLine> lines;

  /// Returns the cell for [digit] (1..9), or null when not present.
  PsychomatrixCell? cellFor(int digit) {
    for (final c in cells) {
      if (c.digit == digit) return c;
    }
    return null;
  }

  @override
  List<Object?> get props => [birthDate, workingNumbers, cells, lines];
}
